lib/temporal.rb

```diff
@@ -81,6 +83,58 @@ def signal_workflow(workflow, signal, workflow_id, run_id, input = nil)
       )
     end
 
+    # Long polls for a workflow to be completed and returns whatever the execute function
+    # returned.
+    # run_id of nil: await the entire workflow completion.  This can span multiple runs
+    # in the case where the workflow uses continue as new.
+    def await_workflow_result(workflow, workflow_id:, run_id: nil, **args)
+      options = args.delete(:options) || {}
+      execution_options = ExecutionOptions.new(workflow, options)
+
+      current_run_id = run_id
+      loop do
+        history_response = client.get_workflow_execution_history(
+          namespace: execution_options.namespace,
+          workflow_id: workflow_id,
+          run_id: current_run_id,
+          wait_for_new_event: true,
+          event_type: :close
+        )
+        event = history_response['history']['events'].first
+        case event.event_type
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED
+          payloads = event['workflow_execution_completed_event_attributes'].result
+          return nil if !payloads # happens when the workflow itself returns nil
+          return Temporal::Client::Converter::Payload::JSON.new.from_payload(payloads['payloads'].first)
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT
+          raise Temporal::WorkflowTimedOut
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED
+          raise Temporal::WorkflowTerminated
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED
+          raise Temporal::WorkflowCanceled
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_FAILED
+          # failure_info: Temporal::Api::Failure::V1::Failure
+          failure_info = event['workflow_execution_failed_event_attributes']['failure']
+          raise Temporal::WorkflowFailed.new(
+            failure_info['message'],
+            stack_trace: failure_info['stack_trace']
+          )
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW
+          new_run_id = event['workflow_execution_continued_as_new_event_attributes']['new_execution_run_id']
+          if run_id
+            # If they specified a run ID, we should throw to let them know they're not getting the result
+            # they wanted.  They can re-call on the new run ID if they want.
+            raise Temporal::WorkflowRunContinuedAsNew.new(new_run_id: new_run_id)
+          else
+            current_run_id = new_run_id
+            # await the next run until the workflow is complete.
```

Comment:
As far as I understand a very common use-case for ContinueAsNew is periodic execution that might actually never stop. We might be sending people into an endless loop here

Reply:
We may, it's true, though I would hope that it'd be quite rare for a developer to make that mistake.
In our wrappers at Stripe, we can easily detect that it's a cron-scheduled workflow and barf if somebody awaits it without a run_id.
In stock temporal-ruby, I thought about adding a `timeout` parameter, but that'd involve spinning up a thread and I wasn't sure if it would be better to just let the client do that part.  Still, I'd be happy to add that in a followup PR and then I think that largely solves this.

Reply:
I would much rather not handle that case at all and remove the loop until we find a better way

Reply:
As I integrated the timeouts, I realized you were right.  The loop is too complicated for what we get out of it.